compiler/optimizer/op.go

```diff
@@ -140,22 +141,83 @@ func analyzeCuts(assignments []dag.Assignment, layout order.Layout) order.Layout
 	if key == nil {
 		return order.Nil
 	}
+	// This loop implements a very simple data flow analysis where we
+	// track the known order through the scoreboard.  If on exit, there
+	// is more than one field of known order, the current optimization
+	// framework cannot handle this so we return unknown (order.Nil)
+	// as a conservative stance to prevent any problematic optimizations.
+	// If there is precisely one field of known order, then that is the
+	// layout we return.  In a future version of the optimizer, we will
+	// generalize this scoreboard concept across the flowgraph for a
+	// comprehensive approach to dataflow analysis.  See issue #2756.
+	scoreboard := make(map[string]field.Path)
+	scoreboard[fieldKey(key)] = key
 	for _, a := range assignments {
-		// XXX This logic (from the original parallelize code)
-		// seems to have a bug.  See issue #2663.
-		if !fieldOf(a.RHS).Equal(key) {
-			continue
-		}
 		lhs := fieldOf(a.LHS)
+		rhs := fieldOf(a.RHS)
 		if lhs == nil {
+			// If we cannot statically determine the data flow,
+			// we give up and return unknown.  This is overly
+			// conservative in general and will miss optimization
+			// opportunities, e.g., we could do dependency
+			// analysis of a complex RHS expression.
 			return order.Nil
 		}
-		if lhs.Equal(key) {
-			return layout
+		lhsKey := fieldKey(lhs)
+		if rhs == nil {
+			// If the RHS depends on a well-defined set of fields
+			// (non of which are unambiguous like this.foo[this.bar]),
+			// and if all of such dependencies do not have an order
+			// to preserve, then we can continue along by clearing
+			// the LHS from the scoreboard knowing that is being set
+			// to something that does not have a defined order.
+			dependencies, ok := fieldsOf(a.RHS)
+			if !ok {
+				return order.Nil
+			}
+			for _, d := range dependencies {
+				key := fieldKey(d)
+				if _, ok := scoreboard[key]; ok {
+					// There's a dependency on an ordered
+					// field but we're not sophisticated
+					// enough here to know if this preserves
+					// its order...
+					return order.Nil
+				}
+			}
```

Comment:
I'm having trouble understanding why this is necessary. It seems like deleting `lhsKey` and continuing on should always be correct (if conservative), regardless of what's inside `a.RHS`. Do you have a motivating example?

Reply:
It's not conservative to delete something from the scoreboard because to be conservative you need to know there is just one known ordering constraint because the returned layout can only represent one field.  This was essentially the bug in the previous logic.  This will be clarified with the USE-DEF set dataflow analysis when we do this properly.

Reply:
> This was essentially the bug in the previous logic.

But the tests pass without this!